atomic_reactor/plugin.py

```diff
@@ -181,36 +184,27 @@ def save_plugin_timestamp(self, plugin, timestamp):
     def save_plugin_duration(self, plugin, duration):
         pass
 
-    def run(self, keep_going=False, buildstep_phase=False):
+    def get_plugin_class(self, plugin_name):
+        try:
+            plugin_class = self.plugin_classes[plugin_name]
+        except KeyError:
+            return None
+        return plugin_class
+
+    def get_available_plugins(self, keep_going):
         """
-        run all requested plugins
+        check requested plugins availability
+        and handle missing plugins
 
         :param keep_going: bool, whether to keep going after unexpected
                                  failure (only used for exit plugins)
-        :param buildstep_phase: bool, when True remaining plugins will
-                                not be executed after a plugin completes
-                                (only used for build-step plugins)
+        :return: list of tuples, runnable plugins data
         """
-        failed_msgs = []
-        plugin_successful = False
-        plugin_response = None
+        plugins_data = []
         for plugin_request in self.plugins_conf:
-            plugin_successful = False
-            try:
-                plugin_name = plugin_request['name']
-            except (TypeError, KeyError):
-                msg = "invalid plugin request, no key 'name': %s" % plugin_request
-                exc = None if keep_going else PluginFailedException(msg)
-                self.on_plugin_failed('?', exc)
-                logger.error(msg)
-                if keep_going:
-                    continue
-                raise exc
-
-            plugin_conf = plugin_request.get("args", {})
-            try:
-                plugin_class = self.plugin_classes[plugin_name]
-            except KeyError:
+            plugin_name = plugin_request['name']
+            plugin_class = self.get_plugin_class(plugin_name)
+            if not plugin_class:
                 if plugin_request.get('required', True):
```

Comment:
Can this still fail?  The plugin name and class were already verified.

Reply:
This will fail if the plugins configuration contains a plugin that passes the schema but was not loaded during `atomic_reactor.plugin.load_plugins()` (e.g., a plugin file was removed or there is a typo in some plugin name)

Reply:
Ah, okay. I missed the distinction between self.plugin_classes and the plugins conf from osv3.